### ISSUE TITLE
NaN et overfitting

### ISSUE THREAD

**Issue by Nic0lAme** [Nic0lAme/tipe-ia#2]
Dans certains cas, mais de manière il semblerait peu aléatoire (arrive de manière systématique pour une certaine sauvegarde du NN), la matrice **renvoit un Nan** ce qui enclenche par les gradients que l'ensemble du NN est Nan
Problèmes possibles :

- **Overflow** du double pour la fonction sigmoid / softmax (dans l'**exponentielle**)
- Une **division par zéro** qui adviendrait à un quelconque moment (mais serait un minimum aléatoire, ce qui ne semble pas être le cas)

Mon instinct me dit que ça doit surement concerné le **softmax**

Point possible de résolution / amélioration : 

- Identifier clairement d'où vient le nan (bonne chance mais la _dernière sauvegarde tombe sur Nan systématiquement au bout de 100 itérations donc pas mal pour test)_
- Instaurer dans la retropropagation un système visant à minimiser les weights/bias (fonction de cout de la forme J = ... + |W|), simplement en additionnant aux gradients de weight/poids une matrice d'un même scalaire

La **2é solution permet également d'éviter l'overfitting,** donc semble être, à faible dose, une bonne idée, même si ça ne réglera pas à coups sûr le problème du Nan (qui est quand même très embêtant)

**Comment by Nic0lAme** [comment]
Théoriquement, le NaN ne peut plus apparaitre
Il venait du calcul du softmax, qui pouvait passer par des calculs de 10^300 ou 10^-300, ce qui résultait en des infty
Les calculs d'exponentielle ont été ramené à 0 pour faciliter et minimiser les calculs
